File: app/src/main/java/com/gaohui/nestedrecyclerview/java/adapter/MultiTypeAdapter.java
package com.gaohui.nestedrecyclerview.java.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.gaohui.nestedrecyclerview.R;
import com.gaohui.nestedrecyclerview.java.view.ChildRecyclerView;
import com.gaohui.nestedrecyclerview.java.viewholder.SimpleCategoryViewHolder;
import com.gaohui.nestedrecyclerview.java.viewholder.SimpleTextViewHolder;

import java.util.ArrayList;

public class MultiTypeAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private ArrayList<Object> mDataList;

    public static  final  int TYPE_TEXT = 0;
    public static  final  int TYPE_CATEGORY = 1;

    SimpleCategoryViewHolder mCategoryViewHolder;
    public MultiTypeAdapter(ArrayList<Object> dataList) {
        mDataList = dataList;
    }

    @Override
    public int getItemViewType(int position) {
        if(mDataList.get(position) instanceof String) {
            return TYPE_TEXT;
        } else  {
            return TYPE_CATEGORY;
        }
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        if(viewType == TYPE_TEXT) {
           return new SimpleTextViewHolder(LayoutInflater.from(
                    viewGroup.getContext()
            ).inflate(R.layout.layout_item_text, viewGroup, false));
        } else  {
            SimpleCategoryViewHolder simpleCategoryViewHolder =
                    new SimpleCategoryViewHolder(LayoutInflater.from(viewGroup.getContext()).inflate(
                            R.layout.layout_item_category_default,
                            viewGroup,
                            false
                    ));
            mCategoryViewHolder = simpleCategoryViewHolder;
            return simpleCategoryViewHolder;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int pos) {
        if(viewHolder instanceof SimpleTextViewHolder) {
//            Log.d("gaohui","pos " + pos + mDataList.get(pos) + mDataList.get(pos));
            ( (SimpleTextViewHolder)viewHolder).mTv.setText((String)mDataList.get(pos));
        } else if(viewHolder instanceof SimpleCategoryViewHolder) {
            ((SimpleCategoryViewHolder)viewHolder).bindData(mDataList.get(pos));
        }
    }

    @Override
    public int getItemCount() {
        return mDataList.size();
    }

    public ChildRecyclerView getCurrentChildRecyclerView() {
        if(mCategoryViewHolder != null) {
            return mCategoryViewHolder.getCurrentChildRecyclerView();
        }
        return null;
    }

    public void destroy() {
        if(mCategoryViewHolder != null) {
            mCategoryViewHolder.destroy();
        }

    }
}

File: app/src/main/java/com/gaohui/nestedrecyclerview/adapter/MultiTypeAdapter.kt
package com.gaohui.nestedrecyclerview.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.gaohui.nestedrecyclerview.ChildRecyclerView
import com.gaohui.nestedrecyclerview.R
import com.gaohui.nestedrecyclerview.bean.CategoryBean
import com.gaohui.nestedrecyclerview.holder.SimpleCategoryViewHolder
import com.gaohui.nestedrecyclerview.holder.SimpleTextViewHolder

class MultiTypeAdapter(private val dataSet:ArrayList<Any>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val  TYPE_TEXT = 0
        private const val TYPE_CATEGORY = 1
    }

    private var mCategoryViewHolder:SimpleCategoryViewHolder? = null

    override fun getItemViewType(position: Int): Int {
        return if(dataSet[position] is String) {
            TYPE_TEXT
        } else {
            TYPE_CATEGORY
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if(viewType == TYPE_TEXT) {
            SimpleTextViewHolder(LayoutInflater.from(viewGroup.context).inflate(R.layout.layout_item_text,viewGroup,false))
        } else {
            val categoryViewHolder = SimpleCategoryViewHolder(LayoutInflater.from(viewGroup.context).inflate(R.layout.layout_item_category,viewGroup,false))
            mCategoryViewHolder =  categoryViewHolder
            return categoryViewHolder
        }
    }

    override fun getItemCount(): Int = dataSet.size

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, pos: Int) {
        if(holder is SimpleTextViewHolder) {
            holder.mTv.text = dataSet[pos] as String
        } else if(holder is SimpleCategoryViewHolder){
            holder.bindData(dataSet[pos] as CategoryBean)
        }
    }

    fun getCurrentChildRecyclerView(): ChildRecyclerView? {
        mCategoryViewHolder?.apply {
           return this.getCurrentChildRecyclerView()
        }
        return null
    }

    fun destroy() {
        mCategoryViewHolder?.destroy()

    }


}